template.go

```diff
@@ -1,20 +1,18 @@
 package function
 
-// Template
-type Template struct {
-	// Name (short name) of this template within the repository.
-	// See .Fullname for the calculated field which is the unique primary id.
-	Name string `yaml:"-"` // use filesystem for name, not yaml
+import (
+	"context"
+)
 
-	// Runtime for which this template applies.
-	Runtime string
-
-	// Repository within which this template is contained.  Value is set to the
-	// currently effective name of the repository, which may vary. It is user-
-	// defined when the repository is added, and can be set to "default" when
-	// the client is loaded in single repo mode. I.e. not canonical.
-	Repository string
+type Template interface {
+	Name() string
+	Runtime() string
+	Repository() string
+	Fullname() string
+	Write(ctx context.Context, f *Function) error
+}
 
+type templateConfig struct {
```

Comment:
I think the concept of a `templateConfig` might be complicating what could really just be one object.

If the reason for separating these fields out is such that their fields can be embedded in Repository and Runtime structs, then I have a whole other long-winded reason why that may not be necessary (tl;dr they share fields, but are not the same because there are indeed small diferences at each level).

This might actualy be simpler and easer kept as a single `template` struct, with a few duplicated fields as necessary.

Reply:
>then I have a whole other long-winded reason why that may not be necessary (tl;dr they share fields, but are not the same because there are indeed small diferences at each level).

@lkingland what are those differences?

Reply:
To me it seems that `BuildConfig`, `HealthEndpoints`, `BuildEnvs` and `Invocation` are just defaults for `fn.Function` that can be defined at level of repository runtime, or template.
Is there any objective problem with that?

Reply:
>If the reason for separating these fields out is such that their fields can be embedded in Repository and Runtime structs,

That's exactly what I want to do in subsequent PR.